### PR TITLE
Feature/add permissions tests for controllers

### PR DESCRIPTION
**Pull request by underwater**
Updated the controllers to make use of the HasPermission guard and decorator instead of explicit verification of permissions.  each time.   There were some edge cases, which I've left alone in this initial p/r

- the case DEMO role in updateUserSettings user.controller.ts
- few cases where business rules are also being verified (as per our discussion)
- few cases where multiple permissions are being checked (example: ImportController)  

i suggest to deal with these in a seperate P/R.

### CODE REVIEW

File: apps/api/src/app/account-balance/account-balance.controller.ts [hunk 1]
```diff
@@ -10,8 +10,10 @@ import {
 } from '@nestjs/common';
 import { REQUEST } from '@nestjs/core';
 import { AuthGuard } from '@nestjs/passport';
-import { AccountBalance } from '@prisma/client';
+import { permissions } from '@ghostfolio/common/permissions';
```
Comment (dtslvr) [inline apps/api/src/app/account-balance/account-balance.controller.ts:13]: This import seems to be a duplicate

File: apps/api/src/app/cache/cache.controller.ts [hunk 2]
```diff
@@ -21,19 +15,8 @@ export class CacheController {
 
   @Post('flush')
   @UseGuards(AuthGuard('jwt'))
+  @HasPermission(permissions.accessAdminControl) // permission needed
```
Comment (dtslvr) [inline apps/api/src/app/cache/cache.controller.ts:18]: Why this comment?

File: apps/api/src/app/import/import.controller.ts [hunk 3]
```diff
@@ -35,17 +36,14 @@ export class ImportController {
 
   @Post()
   @UseGuards(AuthGuard('jwt'))
+  @HasPermission(permissions.createAccount)
```
Comment (dtslvr) [inline apps/api/src/app/import/import.controller.ts:39]: I would change this to `permissions.createOrder`

File: apps/api/src/app/import/import.controller.ts [hunk 4]
```diff
@@ -35,17 +36,14 @@ export class ImportController {
...
   public async import(
     @Body() importData: ImportDataDto,
     @Query('dryRun') isDryRun?: boolean
   ): Promise<ImportResponse> {
     if (
-      !hasPermission(
-        this.request.user.permissions,
-        permissions.createAccount
-      ) ||
       !hasPermission(this.request.user.permissions, permissions.createOrder)
```
Comment (dtslvr) [inline apps/api/src/app/import/import.controller.ts:47]: I would change this to `permissions.createAccount`

**Review by dtslvr — CHANGES_REQUESTED** [review]
Great how much boilerplate you can eliminate with that 👍🏻

I have added a few commens.

File: apps/api/src/app/cache/cache.controller.ts (hunk 2 above)
Comment (underwater) [reply apps/api/src/app/cache/cache.controller.ts:18]: forgot to remote it :-)

File: apps/api/src/app/account-balance/account-balance.controller.ts (hunk 1 above)
Comment (underwater) [reply apps/api/src/app/account-balance/account-balance.controller.ts:13]: agree

File: apps/api/src/app/import/import.controller.ts (hunk 3 above)
Comment (underwater) [reply apps/api/src/app/import/import.controller.ts:39]: OK, but then will do the excplicit check on the createAccount (as mentioned previously, I didn't test yet the permission guard / decorator on cases that check multiple permissions. (leaving that for seperate P/R).  OK with you ?

File: apps/api/src/app/import/import.controller.ts (hunk 4 above)
Comment (underwater) [reply apps/api/src/app/import/import.controller.ts:47]: OK, got it :-)

**Comment by underwater** [comment]
Fixed to comply with suggestsin....
Also revisited few controllers, and removed unused constructor injection of RequestWithUser and its dependencies where this was no longer needed. So In controllers like account-balance, queue, benchmark, the "@Inject(REQUEST) private readonly request: RequestWithUser" is no longer needed.

**Review by dtslvr — CHANGES_REQUESTED** [review]
Thank you for the changes, @underwater.
I've tried it, however I always get `403 Forbidden` when I open pages of the _Admin Control_ panel. Other endpoints (non-admin) seems to work, though. Any idea?

**Comment by underwater** [comment]
Let me check this tomorrow and revert

**Comment by dtslvr** [comment]
> Let me check this tomorrow and revert

Thanks. Do not revert if possible 🙂

**Comment by underwater** [comment]
:-) I meant, check and get back to you :-)

**Comment by underwater** [comment]
I suspect to have found the issue

I was registering HasPermissionGuard globally which means it runs earlier in the pipeline than the AuthGuard which is being applied explicitly to each controoler endpoint ... 

As a result HasPermissionGuard was receiving the request before the AuthGuard has had a chance to populate the 'user' object and the HasPermissionGuard was seeing 'user' object as 'undefined'.

To fix this, I would suggest I skip the global registration of the HasPermissionGuard (which I was doing in the app.module.ts) and instead just include it in our @UseGuards(...) like this @UseGuards(AuthGuard('jwt'), HasPermissionGuard)


So to summarize,
1) controller endpoints that do check permissions and indicate the required permissions using @HasPermissionDecorator will need to use @UseGuards(AuthGuard('jwt'), HasPermissionGuard)

while 

2) controller endpoints that don't check permissions (like getPlatforms()) will continue to use the simple
@UseGuards(AuthGuard('jwt'))


To keep things consistent and to avoid confusion, we could also always use the 
@UseGuards(AuthGuard('jwt'), HasPermissionGuard) 

This would work for all cases, the HasPermissionsGuard defaults to allowing the request to pass when no required permission is specified using the @HasPermission(....) 

let me know what is preferred and I will do the chance and push for you to have a look.

**Comment by underwater** [comment]
We could eventually also consider registering both guards (AuthGuard, and HasPermissionGuard) globally and in the right order, to avoid having to think about this at each endpoint

**Comment by dtslvr** [comment]
Using
```
@UseGuards(AuthGuard('jwt'), HasPermissionGuard)
```
is also fine. To ensure that it is forgotten less often, it may make sense to always use both of them.

**Comment by underwater** [comment]
sure, I can update my p/r to make use of @UseGuards(AuthGuard('jwt'), HasPermissionGuard) everywhere for consistency...

**Comment by underwater** [comment]
> Using
> 
> ```
> @UseGuards(AuthGuard('jwt'), HasPermissionGuard)
> ```
> 
> is also fine. To ensure that it is forgotten less often, it may make sense to always use both of them.

done ! pushed update

**Review by dtslvr — APPROVED** [review]
Looks fine now. Thanks @underwater 🙂

**Pull request merged**